Framework/Algorithms/src/RemoveExpDecay.cpp

```diff
@@ -66,134 +66,93 @@ void MuonRemoveExpDecay::exec() {
-    for (int i = 0; i < specLength; ++i) {
-      PARALLEL_START_INTERUPT_REGION
-      if (spectra[i] > numSpectra) {
-        g_log.error("Spectra size greater than the number of spectra!");
-        throw std::invalid_argument(
-            "Spectra size greater than the number of spectra!");
-      }
-      // Get references to the x data
-      const auto &xIn = inputWS->x(spectra[i]);
-      auto &yOut = outputWS->mutableY(spectra[i]);
-      auto &eOut = outputWS->mutableE(spectra[i]);
-
-      removeDecayData(xIn, inputWS->y(spectra[i]), yOut);
-      removeDecayError(xIn, inputWS->e(spectra[i]), eOut);
-
-      const double normConst = calNormalisationConst(outputWS, spectra[i]);
+    // Remove decay from Y and E
+    outputWS->setHistogram(specNum, removeDecay(inputWS->histogram(specNum)));
 
-      // do scaling and substract by minus 1.0
-      const size_t nbins = outputWS->y(i).size();
-      for (size_t j = 0; j < nbins; j++) {
-        yOut[j] /= normConst;
-        yOut[j] -= 1.0;
-        eOut[j] /= normConst;
-      }
+    // do scaling and subtract 1
+    const double normConst = calNormalisationConst(outputWS, spectra[i]);
+    outputWS->mutableY(specNum) /= normConst;
+    outputWS->mutableY(specNum) -= 1.0;
+    outputWS->mutableE(specNum) /= normConst;
 
-      prog.report();
-      PARALLEL_END_INTERUPT_REGION
-    }
-    PARALLEL_CHECK_INTERUPT_REGION
+    prog.report();
+    PARALLEL_END_INTERUPT_REGION
   }
+  PARALLEL_CHECK_INTERUPT_REGION
 
   // Update Y axis units
   outputWS->setYUnit("Asymmetry");
 
   setProperty("OutputWorkspace", outputWS);
 }
 
-/** This method corrects the errors for one spectrum.
- *	 The muon lifetime is in microseconds not seconds, i.e. 2.1969811 rather
- *   than 0.0000021969811.
- *   This is because the data is in microseconds.
- *   @param inX ::  The X vector
- *   @param inE ::  The input error vector
- *   @param outE :: The output error vector
+/**
+ * Corrects the data and errors for one spectrum.
+ * The muon lifetime is in microseconds, not seconds, because the data is in
+ * microseconds.
+ * @param histogram :: [input] Input histogram
+ * @returns :: Histogram with exponential removed from Y and E
  */
-void MuonRemoveExpDecay::removeDecayError(const HistogramData::HistogramX &inX,
-                                          const HistogramData::HistogramE &inE,
-                                          HistogramData::HistogramE &outE) {
-  // Do the removal
-  for (size_t i = 0; i < inE.size(); ++i) {
-    if (inE[i] != 0.0)
-      outE[i] = inE[i] * exp(inX[i] / MUON_LIFETIME_MICROSECONDS);
-    else
-      outE[i] = exp(inX[i] / MUON_LIFETIME_MICROSECONDS);
-  }
-}
+HistogramData::Histogram MuonRemoveExpDecay::removeDecay(
+    const HistogramData::Histogram &histogram) const {
+  HistogramData::Histogram result(histogram);
+
+  auto &yData = result.mutableY();
+  auto &eData = result.mutableE();
+  assert(yData.size() == eData.size());
```

Comment:
You do not need this assert, `Histogram` guarantees that this is always the case.